### PR TITLE
GEODE-7157: SSLConfig&SSLConf..Factory thread-safe

### PR DESCRIPTION
**Pull request by albertogpz**
SSLConfig and SSLConfigFactory have been changed to be
tread-safe by means of the following:

* SSLConfig has been converted into an immutable class

* The registeredSSLConfig HashMap member of SSLConfigurationFactory
  has been changed into a ConcurrentHashMap and accesses to this
  map other than put and get has been synchronized.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/net/SSLConfigurationFactory.java [hunk 1]
```diff
@@ -82,7 +83,7 @@ public static SSLConfig getSSLConfigForComponent(
     return sslConfig;
   }
 
-  private synchronized void registeredSSLConfigForComponent(
+  private void registeredSSLConfigForComponent(
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/net/SSLConfigurationFactory.java]: Just curious, why is the synchronized removed here, but then added for clearSSLConfigForAllComponents()?  Are we only worried about protecting multiple concurrent clears?

Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/internal/net/SSLConfigurationFactory.java]: Given that the registeredSSLConfig member is now a ConcurrentHashMap it is not required to synchronize the puts (performed on the registeredSSLConfigForComponent method). The clear nevertheless was not protected against concurrent invocations and now it is.

File: geode-core/src/main/java/org/apache/geode/internal/admin/SSLConfig.java [hunk 2]
```diff
@@ -23,110 +23,111 @@
...
+import org.apache.geode.annotations.Immutable;
 import org.apache.geode.distributed.internal.DistributionConfig;
 import org.apache.geode.internal.security.SecurableCommunicationChannel;
 import org.apache.geode.management.internal.SSLUtil;
 
 /**
  * The SSL configuration settings for a GemFire distributed system.
+ *
+ * This class is immutable and the way to construct it is by means of the
+ * the setters plus the build() method of its inner {@link SSLConfig.Build} class.
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/internal/admin/SSLConfig.java]: ok

File: geode-core/src/main/java/org/apache/geode/internal/admin/SSLConfig.java [hunk 3]
```diff
@@ -23,110 +23,111 @@
...
+  private final String alias;
+  @Immutable
+  private final SecurableCommunicationChannel securableCommunicationChannel;
 
   /**
    * SSL implementation-specific key-value pairs. Each key should be prefixed with
    * <code>javax.net.ssl.</code>
    */
+  @Immutable
   private Properties properties = new Properties();
```
[the remark this answers is not in the document]
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/internal/admin/SSLConfig.java]: Yes. You are right. Thanks

File: geode-core/src/main/java/org/apache/geode/internal/net/SSLConfigurationFactory.java [hunk 4]
```diff
@@ -358,25 +365,24 @@ public static void close() {
     getInstance().distributionConfig = null;
   }
 
-  private void clearSSLConfigForAllComponents() {
+  private synchronized void clearSSLConfigForAllComponents() {
```
[the remark this answers is not in the document]
Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/net/SSLConfigurationFactory.java]: Looks like this was resolved by the method being deleted.

[the remark this answers is not in the document]
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/internal/net/SSLConfigurationFactory.java]: Do you still think it is worth to add a ConcurrentTestRunner test given that now SSLConfig is immutable and SSLConfigFactory only has the static instance member?

**Comment by dschneider-pivotal** [comment]
I think it is still worth adding a test that demonstrates that this class is now thread safe.
You have made it so by making it immutable but the test would prove that it actually is immutable.
And the test could catch future regressions if someone breaks the immutable contract.

**Comment by albertogpz** [comment]
> I think it is still worth adding a test that demonstrates that this class is now thread safe.
> You have made it so by making it immutable but the test would prove that it actually is immutable.
> And the test could catch future regressions if someone breaks the immutable contract.

I have been thinking about what test(s) I could add but have not come up with any idea.

Regarding SSLConfig, after the changes it does not offer any public method that would allow to modify its state. It only provides the constructor and getter methods so I cannot think of a test that would assure that in the future the class will remain immutable.

Regarding SSLConfigurationFactory, having removed the deprecated method that returned shared SSLConfig objects (stored in the static HashMap member), now it always returns a new SSLConfig object each time the getSSLConfigForComponent method is invoked.
In this case, the only test case I can think of is one that makes sure that invocations of these methods with the same parameters return different objects. Would this be useful?

I'd appreciate any insight on what tests to add.

Thanks in advance.

**Comment by dschneider-pivotal** [comment]
I you went back to the initial condition (before this pr's work) wouldn't you be able to write a concurrent test that demonstrates that these classes are not thread safe? If so, then that test should start passing with this pr's changes.

**Comment by albertogpz** [comment]
> I you went back to the initial condition (before this pr's work) wouldn't you be able to write a concurrent test that demonstrates that these classes are not thread safe? If so, then that test should start passing with this pr's changes.

The fact is that the method that had thread-safety issues (public static SSLConfig getSSLConfigForComponent(SecurableCommunicationChannel sslEnabledComponent)) has been eliminated.
Therefore, I could use in a test prior to the PR to see failures but I could not in the new version as it would not be present.

**Comment by kohlmu-pivotal** [comment]
I know that I am VERY late into this discussion... I think the essence of WHY has been lost be this refactor. Whilst approving code on correctness of code is good, but context is lacking and thus we have now added a new potential performance issue. Given that there is no documentation as to what the constraints are/were initially this fix is good.

But to fill in the gaps in context:

1. `SSLConfigurationFactory` was supposed to cache `SSLConfig` objects, for performance reasons. This has now been lost. In addition we are now adding more time in processing the `DistributionConfig` EVERY time we need `SSLConfig` for a `SecurableChannel`
1. The `DistributionConfig` stored on the `SSLConfigurationFactory` was really to tie the "cached" `SSLConfig` objects to a `DistributionConfig`
1. `SSLConfigurationFactory` was never supposed to be static. At the time we had to do it, because we could not find a good "home" for it. It should really be an instance that takes a `DistributionConfig` and provides the `SSLConfig` for that config. Not an ad-hoc.
1. Thread safety is really less of a concern, given that the `DistributionConfig` should never change throughout the lifetime of the `DistributedSystem`. If you happen to have 10 threads asking for `SSLConfig` for the `SERVER` at the same time, YES, you could have 10 threads recreating the same 10 `SSLConfig`urations, BUT, they would contain the SAME information. In the end you will 1 cached copy of `SSLConfig`.

I think DO like the new `SSLConfig.Builder` pattern applied now. Immutability of the `SSLConfig` object is great! Maybe, if we can couple, it would "caching" we might be able to offset the performance hit we would now have.

It is also interesting to me that we annotate `final` properties with `Immutable`.

**Comment by albertogpz** [comment]
> I know that I am VERY late into this discussion... I think the essence of WHY has been lost be this refactor. Whilst approving code on correctness of code is good, but context is lacking and thus we have now added a new potential performance issue. Given that there is no documentation as to what the constraints are/were initially this fix is good.
> 
> But to fill in the gaps in context:
> 
>     1. `SSLConfigurationFactory` was supposed to cache `SSLConfig` objects, for performance reasons. This has now been lost. In addition we are now adding more time in processing the `DistributionConfig` EVERY time we need `SSLConfig` for a `SecurableChannel`
> 
>     2. The `DistributionConfig` stored on the `SSLConfigurationFactory` was really to tie the "cached" `SSLConfig` objects to a `DistributionConfig`
> 
>     3. `SSLConfigurationFactory` was never supposed to be static. At the time we had to do it, because we could not find a good "home" for it. It should really be an instance that takes a `DistributionConfig` and provides the `SSLConfig` for that config. Not an ad-hoc.
> 
>     4. Thread safety is really less of a concern, given that the `DistributionConfig` should never change throughout the lifetime of the `DistributedSystem`. If you happen to have 10 threads asking for `SSLConfig` for the `SERVER` at the same time, YES, you could have 10 threads recreating the same 10 `SSLConfig`urations, BUT, they would contain the SAME information. In the end you will 1 cached copy of `SSLConfig`.
> 
> 
> I think DO like the new `SSLConfig.Builder` pattern applied now. Immutability of the `SSLConfig` object is great! Maybe, if we can couple, it would "caching" we might be able to offset the performance hit we would now have.
> 
> It is also interesting to me that we annotate `final` properties with `Immutable`.

Hi Udo,
I noticed that this class was providing a cache for SSLConfig objects. Nevertheless, the method that was making use of the cache appeared as deprecated since Geode 1.3. In the comment, about the deprecation, it was recommended to use another method of the class that created an SSLConfig each time.
Given that the cache had some thread-safety issues to be solved and given that the method managing the was deprecated I figured the best option would be to just remove it.
Anyway, if it was a mistake to remove it, I offer myself to put it back.

**Comment by kohlmu-pivotal** [comment]
@albertogpz Thank you for this.
I think there are some good fixes in the PR. I think we need the `SSLConfig` cache back.. but it needs to be better managed. Possibly there is some stale state and possibly some better approaches to some of the deficiencies.

I had not noticed the deprecation on the method. Possibly another commit that did not take into account the "caching".

Tbh, let's see how this evolves, maybe "cache vs not cached" is a non-event given that we don't really cycle connections that often.

Thank you for the PR though. Great work!

**Pull request merged**